### PR TITLE
[None][feat] Add MTP>1 support for DS-v3.2

### PR DESCRIPTION
**Pull request by lfr-0531**
## Summary by CodeRabbit

* **New Features**
  * Added support for multi-draft token generation in speculative decoding workflows
  * Optimized attention computation for efficient handling of multiple draft tokens during inference
  * Maintains full backward compatibility with existing single-draft configurations

## Description

In the DeepSeek-v3.2 model, the `deep_gemm.fp8_paged_mqa_logits` kernel cannot support next_n > 2, i.e., MTP>1. In this PR, we flatten `q_decode` along the batch_size and seq_len dimensions for MTP > 1 cases. Then we can use the `fp8_paged_mqa_logits` kernel to calculate the MQA. For the causal mask, we can apply a mask to the logits before the top-k. 

Accuracy with NVFP4 model and MTP=3:
```
[TRT-LLM] [I] lm-eval gpqa_diamond_cot_zeroshot_aa results (scores normalized to range 0~100):
|           Tasks            |Version|   Filter   |n-shot|  Metric   |   | Value |   |Stderr|
|----------------------------|------:|------------|-----:|-----------|---|------:|---|-----:|
|gpqa_diamond_cot_zeroshot_aa|      1|strict-match|     0|exact_match|↑  |81.8182|±  | 2.748|

[TRT-LLM] [I] lm-eval gpqa_diamond_cot_zeroshot_aa average accuracy: 81.82
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 1]
```diff
@@ -534,6 +568,39 @@ def prepare(self):
         else:
             self.max_gen_seq_len = 0
 
+        # Because the fp8_paged_mqa_logits only supports seq_len == 1 or 2, so it cannot support
+        # MTP > 1. To haddle this, when MTP > 1, we flatten the q tensor and expand the kv_lens and
```
Comment (chang-l) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Typo:  `haddle` --> `handle`

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 2]
```diff
@@ -534,6 +568,39 @@ def prepare(self):
...
+            self.kv_lens_expanded_host[:num_tokens].copy_(gen_kv_lens_expanded)
+            self.kv_lens_expanded_cuda[:num_tokens].copy_(
+                self.kv_lens_expanded_host[:num_tokens], non_blocking=True)
+
+            # Expand indexer_k_cache_block_offsets (only generation)
+            if self.kv_cache_manager is not None:
+                block_ids = self.kv_cache_manager.get_batch_cache_indices(
+                    self.request_ids)
+                for i in range(self.num_contexts, len(block_ids)):
+                    for j in range(1 + self.max_draft_tokens):
```
Comment (chang-l) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Can we use tensor op instead of nested loops for simplicity?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 3]
```diff
@@ -798,6 +865,15 @@ def prepare(metadata: DSAtrtllmAttentionMetadata):
                 gen_seq_lens, tokens_per_block, metadata.num_sms)
             metadata.scheduler_metadata_buffer.copy_(scheduler_metadata_buffer,
                                                      non_blocking=True)
+            if metadata.max_draft_tokens > 1:
```
[the remark this answers is not in the document]
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Updated. When `metadata.max_draft_tokens <= 1`, prepare `scheduler_metadata_buffer`; otherwise, prepare `scheduler_metadata_buffer_expanded`.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 4]
```diff
@@ -431,6 +431,40 @@ def __post_init__(self):
             dtype=torch.int32,
             capture_graph=capture_graph,
         )
+        self.kv_lens_expanded_cuda = self.get_empty(
```
[the remark this answers is not in the document]
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Added.

**Review by chang-l — APPROVED** [review]

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 5]
```diff
@@ -1067,7 +1069,7 @@ def update_spec_dec_param(
         max_draft_tokens,
         spec_decoding_tensor: Optional['SpecDecodingTensor'] = None,
     ):
-
+        self.max_draft_tokens = max_draft_tokens
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: Move this line to `DSAtrtllmAttentionMetadata`'s `update_spec_dec_param`

**Review by yuxianq — APPROVED** [review]
